src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.cs

```diff
@@ -0,0 +1,499 @@
+                        classType = ClassType.TypeUnsupportedBySourceGen;
+                    }
+                }
+            }
+            else
+            {
+                classType = ClassType.Object;
+
+                if (type.GetConstructor(Type.EmptyTypes) != null && !type.IsAbstract && !type.IsInterface)
+                {
+                    // TODO: support parameterized ctors.
+                    constructionStrategy = ObjectConstructionStrategy.ParameterlessConstructor;
+                }
+
+                Dictionary<string, PropertyMetadata>? ignoredMembers = null;
+
+                for (Type? currentType = type; currentType != null; currentType = currentType.BaseType)
+                {
+                    const BindingFlags bindingFlags =
+                        BindingFlags.Instance |
+                        BindingFlags.Public |
+                        BindingFlags.NonPublic |
+                        BindingFlags.DeclaredOnly;
+
+                    foreach (Reflection.PropertyInfo propertyInfo in currentType.GetProperties(bindingFlags))
+                    {
+                        PropertyMetadata metadata = GetPropertyMetadata(propertyInfo);
+
+                        // Ignore indexers and virtual properties that have overrides that were [JsonIgnore]d.
+                        if (propertyInfo.GetIndexParameters().Length > 0 || PropertyIsOverridenAndIgnored(metadata, ignoredMembers))
+                        {
+                            continue;
+                        }
+
+                        string key = metadata.JsonPropertyName ?? metadata.ClrName;
+
+                        if (metadata.HasGetter || metadata.HasSetter)
+                        {
+                            (propertiesMetadata ??= new()).Add(metadata);
+                        }
+
+                        if (containsOnlyPrimitives && !IsPrimitive(propertyInfo.PropertyType))
+                        {
+                            containsOnlyPrimitives = false;
+                        }
+                    }
+
+                    foreach (FieldInfo fieldInfo in currentType.GetFields(bindingFlags))
+                    {
+                        PropertyMetadata metadata = GetPropertyMetadata(fieldInfo);
+
+                        if (PropertyIsOverridenAndIgnored(metadata, ignoredMembers))
+                        {
+                            continue;
+                        }
+
+                        if (metadata.HasGetter || metadata.HasSetter)
+                        {
+                            (propertiesMetadata ??= new()).Add(metadata);
+                        }
+                    }
+                }
+            }
+
+            string compilableName = type.GetUniqueCompilableTypeName();
+            string friendlyName = type.GetFriendlyTypeName();
+
+            if (_usedFriendlyTypeNames.Contains(friendlyName))
+            {
+                friendlyName = type.GetUniqueFriendlyTypeName();
```

Comment:
Per discussion, we discussed just giving a compile error here which would require the developer to add an attribute to specify the appropriate name to use.

Reply:
Yup, I'll pull this out. I have task to read the value of that attribute property as a follow up to this PR.